lib/vra/catalog_request.rb

```diff
@@ -0,0 +1,131 @@
+      @additional_params.set(key, type, value)
+    end
+
+    def delete_parameter(key)
+      @additional_params.delete(key)
+    end
+
+    def parameters
+      @additional_params.all_entries
+    end
+
+    def subtenant_id
+      @subtenant_id || @catalog_details[:subtenant_id]
+    end
+
+    def validate_params!
+      missing_params = []
+      [ :catalog_id, :cpus, :memory, :requested_for, :subtenant_id ].each do |param|
+        missing_params << param.to_s if send(param).nil?
+      end
+
+      raise ArgumentError, "Unable to submit request, required param(s) missing => #{missing_params.join(', ')}" unless missing_params.empty?
+    end
+
+    def fetch_catalog_item
+      begin
+        response = @client.http_get("/catalog-service/api/consumer/catalogItems/#{@catalog_id}")
+      rescue Vra::Exception::HTTPNotFound
+        raise Vra::Exception::NotFound, "catalog ID #{@catalog_id} does not exist"
+      end
+
+      item = JSON.load(response.body)
+
+      @catalog_details[:tenant_id]    = item['organization']['tenantRef']
+      @catalog_details[:subtenant_id] = item['organization']['subtenantRef']
+      @catalog_details[:blueprint_id] = item['providerBinding']['bindingId']
+    end
+
+    def catalog_detail(key)
+      @catalog_details[key]
+    end
+
+    def request_payload
+      fetch_catalog_item if @catalog_details.empty?
+
+      payload = {
+        '@type' => 'CatalogItemRequest',
+        'catalogItemRef' => {
+          'id' => @catalog_id
+        },
+        'organization' => {
+          'tenantRef'    => catalog_detail(:tenant_id),
+          'subtenantRef' => subtenant_id
+        },
+        'requestedFor' => @requested_for,
+        'state' => 'SUBMITTED',
+        'requestNumber' => 0,
+        'requestData' => {
+          'entries' => [
+            Vra::RequestParameter.new('provider-blueprintId', 'string', catalog_detail(:blueprint_id)).to_h,
+            Vra::RequestParameter.new('provider-provisioningGroupId', 'string', subtenant_id).to_h,
+            Vra::RequestParameter.new('requestedFor', 'string', @requested_for).to_h,
+            Vra::RequestParameter.new('provider-VirtualMachine.CPU.Count', 'integer', @cpus).to_h,
+            Vra::RequestParameter.new('provider-VirtualMachine.Memory.Size', 'integer', @memory).to_h,
+            Vra::RequestParameter.new('provider-VirtualMachine.LeaseDays', 'integer', @lease_days).to_h,
+            Vra::RequestParameter.new('provider-__Notes', 'string', @notes).to_h
+          ]
+        }
+      }
+
+      parameters.each do |entry|
+        payload['requestData']['entries'] << entry.to_h
+      end
+
+      payload
+    end
+
+    def submit
+      validate_params!
+
+      begin
+        response = @client.http_post('/catalog-service/api/consumer/requests', request_payload.to_json)
+      rescue Vra::Exception::HTTPError => e
+        raise Vra::Exception::RequestError, "Unable to submit request: #{e.errors.join(', ')}"
+      rescue
```

Comment:
On the surface, yes.  I did it to specifically put the errors embedded within the API's error response into the exception message rather than getting a generic "500 Server Error" and making the end-user catch the exception and parse/re-raise for themselves.